### PR TITLE
feat(optimism): integrate optimism into Mask ecosystem

### LINKED ISSUE #4868: [Bounty] Integrate Optimism into Mask
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [ ] Intermediate
  - [x] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  [x] 4 weeks

- Bounty Size
  [x] $5000

### Part 1: Problem Description

**Background** Optimism is a Layer 2 scaling solution for Ethereum that can support all of Ethereum's Dapps. Instead of running all computation and data on the Ethereum network, Optimism puts all transaction data on-chain and runs computation off-chain, increasing Ethereum's transactions per second and decreasing transaction fees.

**Solution** As the first step of our cooperation, we, Mask Network, would like to integrate Optimism as one of our default supported blockchain system in our internal Mask Wallet, so that further on-chain interactions would happen directly within Mask Network without leaving the platform. We believe this would further help us attract more users from different projects, aiming at different goals to maximize users experiences in the broad blockchain world.

If you are interested in this bounty, please make sure that you contact us first to see how our structure looks like and proceed to the development afterwards.

### Part 2: Resources

- https://www.optimism.io/
- https://github.com/DimensionDev/Maskbook/tree/master/packages/maskbook/src/plugins/Wallet

### Part 3: Contact and Support

@yisiliu for ideas @Jack-Works for implementations Mask Network discord server: https://discord.gg/4Xccj7a75Q Optimism discord server: https://discord.com/invite/jrnFEvq

### PR DESCRIPTION
**Pull request by developerfred**
## Description

Optimism integration on MaskBook

Closes #4868

## Type of change

- [x] add optimist Layer2 
- [x] add UniswapV3  
- [x] enable UnlockProtocol
- [x] enable dHedge  

## Token  List
- [ ] https://github.com/DimensionDev/Mask-Token-List/pull/49
- [ ] https://github.com/DimensionDev/assets/pull/2/files

## Dex
- [x] Uniswap 
- [x] Juggler.red 

## Mask Assets 
- [x] https://github.com/DimensionDev/assets/pull/2

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

### CODE REVIEW

**Review by Jack-Works — DISMISSED** [review]
cc @guanbinrui

**Comment by guanbinrui** [comment]
Hi @developerfred, we've completed the EVM chain integration document. Would you please read the checklist to ensure you didn't miss anything? Also, if we have something we didn't mention in the document. Let us know it.

Learn more: https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md

**Comment by developerfred** [comment]
> Hi @developerfred, we've completed the EVM chain integration document. Would you please read the checklist to ensure you didn't miss anything? Also, if there is something we didn't mention. Please let us know it.
> 
> Learn more: [`develop`/docs/evm-integration.md](https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md?rgh-link-date=2021-12-27T04%3A16%3A19Z)

OK, I'm going to check the list and do a check up

**Review by Jack-Works — DISMISSED** [review]

**Comment by developerfred** [comment]
> Hi @developerfred, we've completed the EVM chain integration document. Would you please read the checklist to ensure you didn't miss anything? Also, if there is something we didn't mention. Please let us know it.
> 
> Learn more: [`develop`/docs/evm-integration.md](https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md?rgh-link-date=2021-12-27T04%3A16%3A19Z)

done ✅

**Comment by developerfred** [comment]
swap wrapping ✅ 
690d66347...f717ead39 
Screenshot 2022-01-17 at 23 18 21

File: packages/dashboard/src/pages/Settings/api.ts [hunk 1]
```diff
@@ -43,6 +43,11 @@ export const [useCeloNetworkTradeProvider] = createGlobalState(
     Messages.events.celoNetworkTradeProviderSettings.on,
 )
 
+export const [useOpNetworkTradeProvider] = createGlobalState(
```
Comment (UncleBill) [inline packages/dashboard/src/pages/Settings/api.ts]: `useXXTradeProvider` will be removed. see https://github.com/DimensionDev/Maskbook/pull/5516

File: packages/mask/src/extension/background-script/SettingsService.ts [hunk 2]
```diff
@@ -77,9 +78,12 @@ export const [getArbitrumNetworkTradeProvider, setArbitrumNetworkTradeProvider]
 )
 export const [getxDaiNetworkTradeProvider, setxDaiNetworkTradeProvider] = create(xdaiNetworkTradeProviderSettings)
 
+export const [getOpNetworkTradeProvider, setOpNetworkTradeProvider] = create(opNetworkTradeProviderSettings)
```
Comment (UncleBill) [inline packages/mask/src/extension/background-script/SettingsService.ts]: `getXXTraderProvider` will be removed, see https://github.com/DimensionDev/Maskbook/pull/5516

File: packages/mask/src/plugins/Trader/services.ts [hunk 3]
```diff
@@ -45,6 +46,11 @@ currentChainIdSettings.addListener((chainId: ChainId) => {
             if (currentDataProviderSettings.value === DataProvider.UNISWAP_INFO)
                 currentDataProviderSettings.value = DataProvider.COIN_MARKET_CAP
             break
+        case NetworkType.Optimistic:
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/services.ts]: Will be removed, see https://github.com/DimensionDev/Maskbook/pull/5516

File: packages/mask/src/plugins/Trader/settings.ts [hunk 4]
```diff
@@ -88,6 +88,12 @@ export const celoNetworkTradeProviderSettings = createGlobalSettings<TradeProvid
     { primary: () => '' },
 )
 
+export const opNetworkTradeProviderSettings = createGlobalSettings<TradeProvider>(
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/settings.ts]: Will be removed, see https://github.com/DimensionDev/Maskbook/pull/5516

File: packages/web3-constants/evm/coingecko.json [hunk 5]
```diff
@@ -29,9 +32,12 @@
         "BSCT": "",
         "Matic": "matic-network",
         "Mumbai": "arbitrum-one",
-        "Arbitrum": "",
+        "Arbitrum": "ethereum",
```
Comment (UncleBill) [inline packages/web3-constants/evm/coingecko.json]: Please do not make additional changes, make the patch cleaner.

**Review by UncleBill — COMMENTED** [review]
Just mention here that trader settings will be removed (in https://github.com/DimensionDev/Maskbook/pull/5516), next time you merge `origin/develop`, you can drop the changes to trader settings.

File: packages/dashboard/src/pages/Settings/api.ts (hunk 1 above)
Comment (developerfred) [reply packages/dashboard/src/pages/Settings/api.ts]: done

**Comment by UncleBill** [comment]
No. I didn't mean to let you merge fix/remove-settings, please revert the merge. You should merge develop branch only generally.

**Comment by developerfred** [comment]
> No. I didn't mean to let you merge fix/remove-settings, please revert the merge. You should merge develop branch only generally.

I will revert now.

**Review by Jack-Works — DISMISSED** [review]

**Comment by UncleBill** [comment]
Hi, @developerfred , #5516 got merged, you could sync with develop branch and handle the settings conflicts now.

**Comment by developerfred** [comment]
> Hi, @developerfred , #5516 got merged, you could sync with develop branch and handle the settings conflicts now.

done

**Comment by UncleBill** [comment]
Hi, @developerfred , thanks for your updates.

I find out that `Optimistic` doesn't show up in SwapDialog, you ought to test yourself before sending review request/test request.

image

```diff
diff --git a/packages/mask/src/plugins/Trader/base.ts b/packages/mask/src/plugins/Trader/base.ts
index 2c575a7ae..c7b14dc54 100644
--- a/packages/mask/src/plugins/Trader/base.ts
+++ b/packages/mask/src/plugins/Trader/base.ts
@@ -23,6 +23,7 @@ export const base: Plugin.Shared.Definition = {
                     ChainId.Aurora,
                     ChainId.Avalanche,
                     ChainId.Fantom,
+                    ChainId.Optimistic,
                 ],
             },
         },
```

**Review by Jack-Works — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

**Comment by developerfred** [comment]
ref: UniswapV3 Address [uniswap deployer](https://optimistic.etherscan.io/address/0x6c9fc64a53c1b71fb3f9af64d1ae3a4931a5f4e9)

**Review by Jack-Works — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

**Comment by guanbinrui** [comment]
@developerfred Any updates?

**Comment by developerfred** [comment]
@guanbinrui I'm working on rcp interceptor, this is the same problem as boba network sends signed transactions eth_singTransaction is not available on this blockchain

**Comment by guanbinrui** [comment]
> @guanbinrui I'm working on rcp interceptor, this is the same problem as boba network sends signed transactions eth_singTransaction is not available on this blockchain

Do you mean send_rawTransaction? But eth_signTransaction doesn't support on many provider.

**Comment by developerfred** [comment]
> > @guanbinrui I'm working on rcp interceptor, this is the same problem as boba network sends signed transactions eth_singTransaction is not available on this blockchain
> 
> Do you mean send_rawTransaction? But eth_signTransaction doesn't support on many provider.

yes

File: docs/evm-integration.md [hunk 6]
```diff
@@ -55,7 +55,9 @@ If there is no one deployed [these contracts](https://github.com/DimensionDev/mi
 
 - [ ] Multicall Contract
 - [ ] BalanceChecker
-- [ ] Other contracts from the Mask team
+- [ ] [Mask Box](https://github.com/DimensionDev/MysteryBox/tree/add_more_networks)
+- [ ] [Red Packet and NFT Packet](https://github.com/DimensionDev/RedPacket/tree/add_more_networks)
+- [ ] [ITO](https://github.com/DimensionDev/InitialTwitterOffering/tree/add_more_networks)
```
Comment (septs) [inline docs/evm-integration.md]: suggest, create issue and refer issue

because, remote branch may be deleted

**Review by Jack-Works — DISMISSED** [review]

**Review by guanbinrui — DISMISSED** [review]

**Comment by guanbinrui** [comment]
@DimensionDev/qa

**Comment by AinuX** [comment]
@developerfred 

1.  Unable load balance in token list, so can't do further test 

https://user-images.githubusercontent.com/2582974/165907975-b56a0141-84be-4c95-b23a-17dea568033a.mp4

**Comment by developerfred** [comment]
@AinuX Balance checker fixed, I added a proprietary rpc "alchemy rpc" so it can make calls in optimism's rpc api

**Comment by AinuX** [comment]
@developerfred 

Tried with latest build, still can't load balance.

image

**Comment by developerfred** [comment]
@AinuX 

Screenshot 2022-05-26 at 19 55 44
Screenshot 2022-05-26 at 19 56 01


23ecb83c7...e8842947c

**Comment by developerfred** [comment]
Wrap ETH
Screenshot 2022-05-26 at 21 17 02

File: packages/mask/src/plugins/Wallet/apis/debank.ts [hunk 7]
```diff
@@ -31,14 +31,15 @@ export async function getAssetsList(address: string) {
...
+const optimismChainOnDebank = getDeBankConstants(ChainId.Optimism).CHAIN_ID
 const fieldKeys = ['fast', 'normal', 'slow'] as const
 /**
  * Debank's data might be outdated, like gas price for aurora which requires 1 Gwei at least
  * https://twitter.com/AlexAuroraDev/status/1490353255817302016
  * Once debank fixes it, we will remove this modifier.
  */
 function gasModifier(gasDict: GasPriceDictResponse, chain: string) {
-    if ([auroraChainOnDebank, arbitrumChainOnDebank].includes(chain)) {
+    if ([auroraChainOnDebank, arbitrumChainOnDebank, optimismChainOnDebank].includes(chain)) {
```
Comment (UncleBill) [inline packages/mask/src/plugins/Wallet/apis/debank.ts]: Leave only `auroraChainOnDebank`, others don't need to be modified.

File: packages/web3-shared/evm/hooks/useGasPrice.ts [hunk 8]
```diff
@@ -6,9 +6,11 @@ import type { ChainId } from '../types'
 export function useGasPrice(expertedChainId?: ChainId) {
     const currentChainId = useChainId()
     const chainId = expertedChainId ?? currentChainId
+    const optimism = currentChainId === 10
     const web3 = useWeb3({ chainId })
+    const getGasPrice = optimism ? 15000000 : web3.eth.getGasPrice()
 
     return useAsyncRetry(async () => {
-        return web3.eth.getGasPrice()
+        return getGasPrice
```
Comment (UncleBill) [inline packages/web3-shared/evm/hooks/useGasPrice.ts]: Move the logic above inside `useAsyncRetry`

File: packages/web3-shared/evm/hooks/useGasPrice.ts [hunk 9]
```diff
@@ -6,9 +6,11 @@ import type { ChainId } from '../types'
 export function useGasPrice(expertedChainId?: ChainId) {
     const currentChainId = useChainId()
     const chainId = expertedChainId ?? currentChainId
+    const optimism = currentChainId === 10
     const web3 = useWeb3({ chainId })
+    const getGasPrice = optimism ? 15000000 : web3.eth.getGasPrice()
```
Comment (UncleBill) [inline packages/web3-shared/evm/hooks/useGasPrice.ts]: Is gas price for optimism fixed at `15000000`?

**Comment by Jack-Works** [comment]
cc @guanbinrui I think we may help community PR to resolve big merge conflicts comes from the EVM refactor?

**Comment by developerfred** [comment]
@Jack-Works done

File: .i18n-codegen.json [hunk 10]
```diff
@@ -222,6 +222,18 @@
                 "trans": "Translate"
             }
         },
+        {
```
Comment (Jack-Works) [inline .i18n-codegen.json]: Revert changes in this file. They're both duplicated entries

**Review by UncleBill — CHANGES_REQUESTED** [review]

File: packages/web3-shared/evm/hooks/useGasPrice.ts (hunk 9 above)
Comment (developerfred) [reply packages/web3-shared/evm/hooks/useGasPrice.ts]: Yes!

File: .i18n-codegen.json [hunk 11]
```diff
@@ -265,6 +265,30 @@
                 "namespace": "com.maskbook.gitcoin",
                 "trans": "Translate"
             }
+        },
+        {
+            "input": "./packages/mask/src/plugins/Trader/locales/en-US.json",
```
Comment (Jack-Works) [inline .i18n-codegen.json]: revert changes in this file plz

File: packages/web3-shared/evm/constants/chains.json [hunk 12]
```diff
@@ -321,18 +322,59 @@
...
+            {
+                "name": "optimisc",
+                "url": "https://Optimistic.etherscan.io/"
+            }
+        ]
+    },
+    {
+        "chainId": 69,
+        "name": "Optimism Ethereum Testnet Kovan",
+        "chain": "KOR",
```
Comment (guanbinrui) [inline packages/web3-shared/evm/constants/chains.json]:
```suggestion
        "chain": "Optimism",
```

File: packages/web3-shared/evm/constants/chains.json [hunk 13]
```diff
@@ -321,18 +322,59 @@
...
+            "decimals": 18,
+            "logoURI": "https://static.debank.com/image/chain/logo_url/op/01ae734fe781c9c2ae6a4cc7e9244056.png"
+        },
+        "infoURL": "https://optimism.io",
+        "shortName": "kor"
+    },
+    {
+        "chainId": 420,
+        "name": "Optimism Ethereum Testnet Goerli",
+        "chain": "Goerli",
```
Comment (guanbinrui) [inline packages/web3-shared/evm/constants/chains.json]:
```suggestion
        "chain": "Optimism",
```

File: packages/web3-shared/evm/constants/chains.json [hunk 14]
```diff
@@ -321,18 +322,59 @@
     },
     {
         "chainId": 10,
-        "type": "Optimistic",
-        "name": "OPT",
+        "name": "Optimism Ethereum",
```
Comment (guanbinrui) [inline packages/web3-shared/evm/constants/chains.json]:
```suggestion
        "name": "Optimism",
```

File: packages/web3-shared/evm/constants/chains.json [hunk 15]
```diff
@@ -321,18 +322,59 @@
...
+        "explorers": [
+            {
+                "name": "optimisc",
+                "url": "https://Optimistic.etherscan.io/"
+            }
+        ]
+    },
+    {
+        "chainId": 69,
+        "name": "Optimism Ethereum Testnet Kovan",
```
Comment (guanbinrui) [inline packages/web3-shared/evm/constants/chains.json]:
```suggestion
        "name": "Optimism Testnet Kovan",
```

File: packages/web3-shared/evm/constants/chains.json [hunk 16]
```diff
@@ -321,18 +322,59 @@
...
+            "symbol": "KOR",
+            "decimals": 18,
+            "logoURI": "https://static.debank.com/image/chain/logo_url/op/01ae734fe781c9c2ae6a4cc7e9244056.png"
+        },
+        "infoURL": "https://optimism.io",
+        "shortName": "kor"
+    },
+    {
+        "chainId": 420,
+        "name": "Optimism Ethereum Testnet Goerli",
```
Comment (guanbinrui) [inline packages/web3-shared/evm/constants/chains.json]:
```suggestion
        "name": "Optimism Testnet Goerli",
```

File: packages/web3-constants/evm/zerion.json [hunk 17]
```diff
@@ -12,6 +12,9 @@
         "Arbitrum": "arbitrum-assets",
         "Arbitrum_Rinkeby": "",
         "xDai": "xdai-assets",
+        "Optimism": "Optimism-assets",
```
Comment (guanbinrui) [inline packages/web3-constants/evm/zerion.json]:
```suggestion
        "Optimism": "optimism-assets",
```

File: packages/web3-constants/evm/zerion.json [hunk 18]
```diff
@@ -35,6 +38,9 @@
         "Arbitrum": "arbitrum-transactions",
         "Arbitrum_Rinkeby": "",
         "xDai": "xdai-transactions",
+        "Optimism": "Optimism-transactions",
```
Comment (guanbinrui) [inline packages/web3-constants/evm/zerion.json:41]:
```suggestion
        "Optimism": "optimism-transactions",
```

File: packages/web3-constants/evm/token-list.json [hunk 19]
```diff
@@ -12,6 +12,13 @@
         "Arbitrum": ["https://tokens.r2d2.to/latest/42161/tokens.json"],
         "Arbitrum_Rinkeby": ["https://tokens.r2d2.to/latest/421611/tokens.json"],
         "xDai": ["https://tokens.r2d2.to/latest/100/tokens.json"],
+        "Optimism": [
+            "https://static.optimism.io/optimism.tokenlist.json",
+            "https://tokens.r2d2.to/latest/10/tokens.json",
+            "https://token-list.sushi.com"
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token-list.json]:
```suggestion
            "https://tokens.r2d2.to/latest/10/tokens.json"
```

File: packages/web3-constants/evm/rpc.json [hunk 20]
```diff
@@ -83,6 +90,9 @@
         "Arbitrum": [0, 0, 0, 0, 0],
         "Arbitrum_Rinkeby": [0, 0, 0, 0, 0],
         "xDai": [0, 0, 0, 0, 0],
+        "Optimism": [0, 1, 2, 3, 4],
```
Comment (guanbinrui) [inline packages/web3-constants/evm/rpc.json]:
```suggestion
        "Optimism": [0, 1, 2, 2],
```

File: packages/web3-constants/evm/mask-box.json [hunk 21]
```diff
@@ -35,6 +38,9 @@
         "Arbitrum": 0,
         "Arbitrum_Rinkeby": 0,
         "xDai": 0,
+        "Optimism": 2283229,
```
Comment (guanbinrui) [inline packages/web3-constants/evm/mask-box.json]: @UncleBill Please check this.

File: packages/web3-constants/evm/explorer.json [hunk 22]
```diff
@@ -35,6 +38,9 @@
         "Arbitrum": "https://api.arbiscan.io/api",
         "Arbitrum_Rinkeby": "",
         "xDai": "https://blockscout.com/xdai/mainnet/api",
+        "Optimism": "https://api-Optimism.etherscan.io/api",
+        "Optimism_Kovan": "https://kovan-Optimism.etherscan.io",
+        "Optimism_Goerli": "https://goerli-Optimism.etherscan.io",
```
Comment (guanbinrui) [inline packages/web3-constants/evm/explorer.json]:
```suggestion
        "Optimism": "https://api-optimism.etherscan.io/api",
        "Optimism_Kovan": "https://kovan-optimism.etherscan.io",
        "Optimism_Goerli": "https://goerli-optimism.etherscan.io",
```

Comment (guanbinrui) [reply packages/web3-constants/evm/explorer.json]: Does these links work?

File: packages/web3-constants/evm/coingecko.json [hunk 23]
```diff
@@ -35,6 +38,9 @@
         "Arbitrum": "",
         "Arbitrum_Rinkeby": "",
         "xDai": "xdai",
+        "Optimism": "ethereum",
```
Comment (guanbinrui) [inline packages/web3-constants/evm/coingecko.json:41]: This field is wrong.

File: packages/web3-constants/evm/mask-box.json (hunk 21 above)
Comment (UncleBill) [reply packages/web3-constants/evm/mask-box.json]: @developerfred It should be **1272779** instead. https://optimistic.etherscan.io/tx/0xa5fa95b9c2305632ff0f10fed5e1277461f07e49855c2e6b5913279f146b00f5

Please check it from https://github.com/DimensionDev/mysteryBox/#mysterybox--masktestnft.

File: packages/web3-constants/evm/coingecko.json (hunk 23 above)
Comment (guanbinrui) [reply packages/web3-constants/evm/coingecko.json:41]: It's `optimism`, is it?

File: packages/mask/src/plugins/Wallet/services/account.ts [hunk 24]
```diff
@@ -64,6 +64,7 @@ export async function getSupportedNetworks() {
         Flags.polygon_enabled ? NetworkType.Polygon : undefined,
         Flags.arbitrum_enabled ? NetworkType.Arbitrum : undefined,
         Flags.xdai_enabled ? NetworkType.xDai : undefined,
+        Flags.Optimism_enabled ? NetworkType.Optimism : undefined,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/services/account.ts]:
```suggestion
        Flags.optimism_enabled ? NetworkType.Optimism : undefined,
```

File: packages/mask/src/plugins/UnlockProtocol/base.ts [hunk 25]
```diff
@@ -14,7 +14,7 @@ export const base: Plugin.Shared.Definition = {
         target: 'insider',
         web3: {
             [NetworkPluginID.PLUGIN_EVM]: {
-                supportedChainIds: [ChainId.Mainnet, ChainId.xDai, ChainId.Matic, ChainId.Rinkeby],
+                supportedChainIds: [ChainId.Mainnet, ChainId.xDai, ChainId.Matic, ChainId.Rinkeby, ChainId.Optimism],
```
Comment (guanbinrui) [inline packages/mask/src/plugins/UnlockProtocol/base.ts]:
```suggestion
                supportedChainIds: [ChainId.Mainnet, ChainId.xDai, ChainId.Matic, ChainId.Rinkeby],
```

File: packages/mask/src/plugins/Trader/constants/0x.ts [hunk 26]
```diff
@@ -6,14 +6,15 @@ export const ZRX_BASE_URL: Record<NetworkType, string> = {
...
+    [NetworkType.Optimism]: 'https://Optimism.api.0x.org/',
     [NetworkType.Avalanche]: 'https://avalanche.api.0x.org/',
     [NetworkType.Celo]: 'https://celo.api.0x.org/',
     [NetworkType.Fantom]: 'https://fantom.api.0x.org/',
     [NetworkType.Aurora]: 'https://aurora.api.0x.org/',
     [NetworkType.Boba]: '',
     [NetworkType.Fuse]: '',
     [NetworkType.Metis]: '',
-    [NetworkType.Optimistic]: '',
+    [NetworkType.Optimism]: '',
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/constants/0x.ts]: It's a duplicate field.

File: packages/mask/src/plugins/Trader/constants/0x.ts [hunk 27]
```diff
@@ -6,14 +6,15 @@ export const ZRX_BASE_URL: Record<NetworkType, string> = {
     [NetworkType.Polygon]: 'https://polygon.api.0x.org/',
     [NetworkType.Arbitrum]: 'https://arbitrum.api.0x.org/',
     [NetworkType.xDai]: 'https://xdai.api.0x.org/',
+    [NetworkType.Optimism]: 'https://Optimism.api.0x.org/',
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/constants/0x.ts]:
```suggestion
    [NetworkType.Optimism]: 'https://optimism.api.0x.org/',
```

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketInHistoryList.tsx [hunk 28]
```diff
@@ -248,7 +248,7 @@ export function RedPacketInHistoryList(props: RedPacketInHistoryListProps) {
                             <Typography variant="body1" className={classNames(classes.info, classes.message)}>
                                 {t.history_total_amount({
                                     amount: formatBalance(history.total, historyToken?.decimals, 6),
-                                    symbol: historyToken?.symbol,
+                                    symbol: historyToken?.symbol!,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketInHistoryList.tsx]:
```suggestion
                                    symbol: historyToken?.symbol,
```

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketInHistoryList.tsx [hunk 29]
```diff
@@ -328,7 +328,7 @@ export function RedPacketInHistoryList(props: RedPacketInHistoryListProps) {
                                         historyToken?.decimals,
                                         6,
                                     ),
-                                    symbol: historyToken?.symbol,
+                                    symbol: historyToken?.symbol!,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketInHistoryList.tsx]:
```suggestion
                                    symbol: historyToken?.symbol,
```

File: packages/mask/shared/flags.ts [hunk 30]
```diff
@@ -30,6 +30,7 @@ export const Flags = {
     polygon_enabled: true,
     arbitrum_enabled: true,
     xdai_enabled: true,
+    Optimism_enabled: true,
```
Comment (guanbinrui) [inline packages/mask/shared/flags.ts]:
```suggestion
    optimism_enabled: true,
```

File: packages/web3-constants/evm/explorer.json (hunk 22 above)
Comment (developerfred) [reply packages/web3-constants/evm/explorer.json]: correct is: "https://goerli-optimistic.etherscan.io/

File: packages/web3-constants/evm/coingecko.json (hunk 23 above)
Comment (developerfred) [reply packages/web3-constants/evm/coingecko.json:41]: > It's `optimism`, is it?

yes but the coingecko api did not return the price of ethereum to the optimism network, I don't know now if that has changed.

Comment (developerfred) [reply packages/web3-constants/evm/coingecko.json:41]: like ```"Aurora": "ethereum",```

**Comment by AinuX** [comment]
@developerfred 

1. First time to reach Optimism, report an error
image
2. Token icon is incorrect 
image
3. Connect to Optimism, chain icon showed as ETH
image
4. Swap not work, and also did not see Uniswap, Juggler.red in swap provider list.

https://user-images.githubusercontent.com/2582974/177269485-c253d417-beb5-475b-85b5-ad1036460802.mp4

5. Incorrect chain info in twitter share, maybe it's the same root cause like issue 2
image
6. ITO not work.

7. Gas setting not work 
image

8. Fail to transfer token with mask wallet.
https://user-images.githubusercontent.com/2582974/177274360-a42bd443-b2b7-4b43-aa42-8efb7b9355d9.mp4

File: packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts [hunk 31]
```diff
@@ -54,34 +37,34 @@ export function useAllTradeComputed(
...
-        traderEstimateGas: quickSwapEstimateGas,
-    } = useUniswapV2Like(tradeProviders, TradeProvider.QUICKSWAP, inputAmount_, inputToken, outputToken)
-
-    // pancake swap
-    const {
-        trader_: pancakeSwap_,
-        trader: pancakeSwap,
-        traderEstimateGas: pancakeSwapEstimateGas,
-    } = useUniswapV2Like(tradeProviders, TradeProvider.PANCAKESWAP, inputAmount_, inputToken, outputToken)
+    //
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts]: ?

**Comment by developerfred** [comment]
Uhulll @guanbinrui @nuanyang233

**Pull request merged**